File: parent_control.cs
// parent_control.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading;
using System.Runtime.InteropServices;

class ParentControl
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "red" => "\x1b[91m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    class Config
    {
        public Rules Rules { get; set; } = new();
        public bool Active { get; set; }
        public string StartTime { get; set; }
        public int TotalTime { get; set; }
    }

    class Rules
    {
        public List<string> Sites { get; set; } = new();
        public int TimeLimit { get; set; }
        public List<string> Apps { get; set; } = new();
    }

    static string ConfigDir => Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".parent_control");
    static string ConfigFile => Path.Combine(ConfigDir, "config.json");
    static string HostsFile => RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
        ? @"C:\Windows\System32\drivers\etc\hosts"
        : "/etc/hosts";
    const string BLOCK_IP = "127.0.0.1";

    static void EnsureConfig()
    {
        Directory.CreateDirectory(ConfigDir);
        if (!File.Exists(ConfigFile))
        {
            var def = new Config();
            File.WriteAllText(ConfigFile, JsonSerializer.Serialize(def, new JsonSerializerOptions { WriteIndented = true }));
        }
    }

    static Config LoadConfig()
    {
        EnsureConfig();
        return JsonSerializer.Deserialize<Config>(File.ReadAllText(ConfigFile)) ?? new Config();
    }

    static void SaveConfig(Config cfg)
    {
        File.WriteAllText(ConfigFile, JsonSerializer.Serialize(cfg, new JsonSerializerOptions { WriteIndented = true }));
    }

    static List<string> ReadHosts()
    {
        return File.ReadAllLines(HostsFile).ToList();
    }

    static void WriteHosts(List<string> lines)
    {
        File.WriteAllLines(HostsFile, lines);
    }

    static void UpdateHosts(List<string> domains, string action)
    {
        var lines = ReadHosts();
        var newLines = lines.Where(line => !domains.Any(d => line.Trim().StartsWith(BLOCK_IP) && line.Contains(d))).ToList();
        if (action == "add")
        {
            foreach (var d in domains) newLines.Add($"{BLOCK_IP} {d}");
        }
        WriteHosts(newLines);
    }

    static void AddRule(string type, string value)
    {
        var cfg = LoadConfig();
        if (type == "site")
        {
            if (!cfg.Rules.Sites.Contains(value))
            {
                cfg.Rules.Sites.Add(value);
                UpdateHosts(new List<string>{ value }, "add");
            }
            else
            {
                Console.WriteLine(Colorize("Site already in list", "yellow"));
                return;
            }
        }
        else if (type == "time")
        {
            if (!int.TryParse(value, out int mins))
            {
                Console.WriteLine(Colorize("Invalid time value", "red"));
                return;
            }
            cfg.Rules.TimeLimit = mins;
        }
        else if (type == "app")
        {
            if (!cfg.Rules.Apps.Contains(value))
                cfg.Rules.Apps.Add(value);
            else
            {
                Console.WriteLine(Colorize("App already in list", "yellow"));
                return;
            }
        }
        else
        {
            Console.WriteLine(Colorize("Unknown rule type", "red"));
            return;
        }
        SaveConfig(cfg);
        Console.WriteLine(Colorize($"Rule added: {type} -> {value}", "green"));
    }

    static void RemoveRule(string type, string value)
    {
        var cfg = LoadConfig();
        if (type == "site")
        {
            if (cfg.Rules.Sites.Remove(value))
                UpdateHosts(new List<string>{ value }, "remove");
            else
            {
                Console.WriteLine(Colorize("Site not found", "yellow"));
                return;
            }
        }
        else if (type == "time")
        {
            cfg.Rules.TimeLimit = 0;
        }
        else if (type == "app")
        {
            if (!cfg.Rules.Apps.Remove(value))
            {
                Console.WriteLine(Colorize("App not found", "yellow"));
                return;
            }
        }
        else
        {
            Console.WriteLine(Colorize("Unknown rule type", "red"));
            return;
        }
        SaveConfig(cfg);
        Console.WriteLine(Colorize($"Rule removed: {type} -> {value}", "green"));
    }

    static void ListRules()
    {
        var cfg = LoadConfig();
        Console.WriteLine(Colorize("Current rules:", "blue"));
        if (cfg.Rules.Sites.Any())
        {
            Console.WriteLine(Colorize("  Blocked sites:", "yellow"));
            foreach (var s in cfg.Rules.Sites) Console.WriteLine($"    - {s}");
        }
        if (cfg.Rules.TimeLimit > 0)
            Console.WriteLine(Colorize($"  Time limit: {cfg.Rules.TimeLimit} min", "yellow"));
        if (cfg.Rules.Apps.Any())
        {
            Console.WriteLine(Colorize("  Blocked apps:", "yellow"));
            foreach (var a in cfg.Rules.Apps) Console.WriteLine($"    - {a}");
        }
        if (!cfg.Rules.Sites.Any() && cfg.Rules.TimeLimit == 0 && !cfg.Rules.Apps.Any())
            Console.WriteLine(Colorize("  No rules.", "yellow"));
    }

    static void EnableControl()
    {
        var cfg = LoadConfig();
        if (cfg.Active)
        {
            Console.WriteLine(Colorize("Parent control already enabled.", "yellow"));
            return;
        }
        cfg.Active = true;
        cfg.StartTime = DateTime.Now.ToString("o");
        SaveConfig(cfg);
        if (cfg.Rules.Sites.Any()) UpdateHosts(cfg.Rules.Sites, "add");
        Console.WriteLine(Colorize("Parent control enabled.", "green"));
    }

    static void DisableControl()
    {
        var cfg = LoadConfig();
        if (!cfg.Active)
        {
            Console.WriteLine(Colorize("Parent control already disabled.", "yellow"));
            return;
        }
        cfg.Active = false;
        if (cfg.Rules.Sites.Any()) UpdateHosts(cfg.Rules.Sites, "remove");
        SaveConfig(cfg);
        Console.WriteLine(Colorize("Parent control disabled.", "green"));
    }

    static void ShowStatus()
    {
        var cfg = LoadConfig();
        Console.WriteLine(Colorize($"Parent control: {(cfg.Active ? "ENABLED" : "DISABLED")}", "blue"));
        if (cfg.Active && !string.IsNullOrEmpty(cfg.StartTime))
        {
            var start = DateTime.Parse(cfg.StartTime);
            var elapsed = (int)(DateTime.Now - start).TotalMinutes;
            Console.WriteLine(Colorize($"Session time: {elapsed} min", "yellow"));
            if (cfg.Rules.TimeLimit > 0)
                Console.WriteLine(Colorize($"Time limit: {cfg.Rules.TimeLimit} min", "yellow"));
        }
        Console.WriteLine(Colorize($"Blocked sites: {cfg.Rules.Sites.Count}", "yellow"));
        Console.WriteLine(Colorize($"Blocked apps: {cfg.Rules.Apps.Count}", "yellow"));
    }

    static void GenerateReport()
    {
        var cfg = LoadConfig();
        Console.WriteLine(Colorize("Activity report:", "blue"));
        Console.WriteLine($"  Date: {DateTime.Now:yyyy-MM-dd}");
        Console.WriteLine($"  Total time: {cfg.TotalTime} min");
        Console.WriteLine($"  Time limit: {cfg.Rules.TimeLimit} min");
        Console.WriteLine($"  Blocked sites: {cfg.Rules.Sites.Count}");
        Console.WriteLine($"  Blocked apps: {cfg.Rules.Apps.Count}");
    }

    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine(Colorize("Usage: parent_control add|remove|list|enable|disable|status|report [args...]", "yellow"));
            return;
        }
        string cmd = args[0];
        var rest = args.Skip(1).ToList();

        switch (cmd)
        {
            case "add":
                if (rest.Count < 2) { Console.WriteLine(Colorize("Usage: add <site|time|app> <value>", "yellow")); return; }
                AddRule(rest[0], rest[1]);
                break;
            case "remove":
                if (rest.Count < 2) { Console.WriteLine(Colorize("Usage: remove <site|time|app> <value>", "yellow")); return; }
                RemoveRule(rest[0], rest[1]);
                break;
            case "list":
                ListRules();
                break;
            case "enable":
                EnableControl();
                break;
            case "disable":
                DisableControl();
                break;
            case "status":
                ShowStatus();
                break;
            case "report":
                GenerateReport();
                break;
            default:
                Console.WriteLine(Colorize($"Unknown command: {cmd}", "red"));
                break;
        }
    }
}

File: parent_control.java
// parent_control.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.time.format.*;
import java.util.*;
import java.util.stream.*;
import com.google.gson.*;

public class parent_control {
    private static final String RESET = "\u001B[0m";
    private static final String GREEN = "\u001B[92m";
    private static final String RED = "\u001B[91m";
    private static final String YELLOW = "\u001B[93m";
    private static final String BLUE = "\u001B[94m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static class Rules {
        List<String> sites = new ArrayList<>();
        int timeLimit;
        List<String> apps = new ArrayList<>();
    }

    private static class Config {
        Rules rules = new Rules();
        boolean active;
        String startTime;
        int totalTime;
    }

    private static String configDir = System.getProperty("user.home") + "/.parent_control";
    private static String configFile = configDir + "/config.json";
    private static String hostsFile = System.getProperty("os.name").toLowerCase().contains("win")
        ? "C:\\Windows\\System32\\drivers\\etc\\hosts"
        : "/etc/hosts";
    private static final String BLOCK_IP = "127.0.0.1";

    private static void ensureConfig() throws IOException {
        Files.createDirectories(Paths.get(configDir));
        if (!Files.exists(Paths.get(configFile))) {
            Config def = new Config();
            saveConfig(def);
        }
    }

    private static Config loadConfig() throws IOException {
        ensureConfig();
        String json = new String(Files.readAllBytes(Paths.get(configFile)));
        Gson gson = new Gson();
        return gson.fromJson(json, Config.class);
    }

    private static void saveConfig(Config cfg) throws IOException {
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        String json = gson.toJson(cfg);
        Files.write(Paths.get(configFile), json.getBytes());
    }

    private static List<String> readHosts() throws IOException {
        return Files.readAllLines(Paths.get(hostsFile));
    }

    private static void writeHosts(List<String> lines) throws IOException {
        Files.write(Paths.get(hostsFile), lines);
    }

    private static void updateHosts(List<String> domains, String action) throws IOException {
        List<String> lines = readHosts();
        List<String> newLines = lines.stream()
            .filter(line -> domains.stream().noneMatch(d -> line.trim().startsWith(BLOCK_IP) && line.contains(d)))
            .collect(Collectors.toList());
        if (action.equals("add")) {
            for (String d : domains) newLines.add(BLOCK_IP + " " + d);
        }
        writeHosts(newLines);
    }

    private static void addRule(String type, String value) throws IOException {
        Config cfg = loadConfig();
        if (type.equals("site")) {
            if (!cfg.rules.sites.contains(value)) {
                cfg.rules.sites.add(value);
                updateHosts(List.of(value), "add");
            } else {
                System.out.println(colorize("Site already in list", YELLOW));
                return;
            }
        } else if (type.equals("time")) {
            int mins;
            try { mins = Integer.parseInt(value); } catch (NumberFormatException e) {
                System.out.println(colorize("Invalid time value", RED));
                return;
            }
            cfg.rules.timeLimit = mins;
        } else if (type.equals("app")) {
            if (!cfg.rules.apps.contains(value)) {
                cfg.rules.apps.add(value);
            } else {
                System.out.println(colorize("App already in list", YELLOW));
                return;
            }
        } else {
            System.out.println(colorize("Unknown rule type", RED));
            return;
        }
        saveConfig(cfg);
        System.out.println(colorize("Rule added: " + type + " -> " + value, GREEN));
    }

    private static void removeRule(String type, String value) throws IOException {
        Config cfg = loadConfig();
        if (type.equals("site")) {
            if (cfg.rules.sites.remove(value)) {
                updateHosts(List.of(value), "remove");
            } else {
                System.out.println(colorize("Site not found", YELLOW));
                return;
            }
        } else if (type.equals("time")) {
            cfg.rules.timeLimit = 0;
        } else if (type.equals("app")) {
            if (!cfg.rules.apps.remove(value)) {
                System.out.println(colorize("App not found", YELLOW));
                return;
            }
        } else {
            System.out.println(colorize("Unknown rule type", RED));
            return;
        }
        saveConfig(cfg);
        System.out.println(colorize("Rule removed: " + type + " -> " + value, GREEN));
    }

    private static void listRules() throws IOException {
        Config cfg = loadConfig();
        System.out.println(colorize("Current rules:", BLUE));
        if (!cfg.rules.sites.isEmpty()) {
            System.out.println(colorize("  Blocked sites:", YELLOW));
            for (String s : cfg.rules.sites) System.out.println("    - " + s);
        }
        if (cfg.rules.timeLimit > 0)
            System.out.println(colorize("  Time limit: " + cfg.rules.timeLimit + " min", YELLOW));
        if (!cfg.rules.apps.isEmpty()) {
            System.out.println(colorize("  Blocked apps:", YELLOW));
            for (String a : cfg.rules.apps) System.out.println("    - " + a);
        }
        if (cfg.rules.sites.isEmpty() && cfg.rules.timeLimit == 0 && cfg.rules.apps.isEmpty())
            System.out.println(colorize("  No rules.", YELLOW));
    }

    private static void enableControl() throws IOException {
        Config cfg = loadConfig();
        if (cfg.active) {
            System.out.println(colorize("Parent control already enabled.", YELLOW));
            return;
        }
        cfg.active = true;
        cfg.startTime = Instant.now().toString();
        saveConfig(cfg);
        if (!cfg.rules.sites.isEmpty()) updateHosts(cfg.rules.sites, "add");
        System.out.println(colorize("Parent control enabled.", GREEN));
    }

    private static void disableControl() throws IOException {
        Config cfg = loadConfig();
        if (!cfg.active) {
            System.out.println(colorize("Parent control already disabled.", YELLOW));
            return;
        }
        cfg.active = false;
        if (!cfg.rules.sites.isEmpty()) updateHosts(cfg.rules.sites, "remove");
        saveConfig(cfg);
        System.out.println(colorize("Parent control disabled.", GREEN));
    }

    private static void showStatus() throws IOException {
        Config cfg = loadConfig();
        System.out.println(colorize("Parent control: " + (cfg.active ? "ENABLED" : "DISABLED"), BLUE));
        if (cfg.active && cfg.startTime != null) {
            Instant start = Instant.parse(cfg.startTime);
            long elapsed = Duration.between(start, Instant.now()).toMinutes();
            System.out.println(colorize("Session time: " + elapsed + " min", YELLOW));
            if (cfg.rules.timeLimit > 0)
                System.out.println(colorize("Time limit: " + cfg.rules.timeLimit + " min", YELLOW));
        }
        System.out.println(colorize("Blocked sites: " + cfg.rules.sites.size(), YELLOW));
        System.out.println(colorize("Blocked apps: " + cfg.rules.apps.size(), YELLOW));
    }

    private static void generateReport() throws IOException {
        Config cfg = loadConfig();
        System.out.println(colorize("Activity report:", BLUE));
        System.out.println("  Date: " + LocalDate.now());
        System.out.println("  Total time: " + cfg.totalTime + " min");
        System.out.println("  Time limit: " + cfg.rules.timeLimit + " min");
        System.out.println("  Blocked sites: " + cfg.rules.sites.size());
        System.out.println("  Blocked apps: " + cfg.rules.apps.size());
    }

    public static void main(String[] args) throws IOException {
        if (args.length < 1) {
            System.out.println(colorize("Usage: parent_control add|remove|list|enable|disable|status|report [args...]", YELLOW));
            return;
        }
        String cmd = args[0];
        List<String> rest = Arrays.asList(args).subList(1, args.length);

        switch (cmd) {
            case "add":
                if (rest.size() < 2) { System.out.println(colorize("Usage: add <site|time|app> <value>", YELLOW)); return; }
                addRule(rest.get(0), rest.get(1));
                break;
            case "remove":
                if (rest.size() < 2) { System.out.println(colorize("Usage: remove <site|time|app> <value>", YELLOW)); return; }
                removeRule(rest.get(0), rest.get(1));
                break;
            case "list":
                listRules();
                break;
            case "enable":
                enableControl();
                break;
            case "disable":
                disableControl();
                break;
            case "status":
                showStatus();
                break;
            case "report":
                generateReport();
                break;
            default:
                System.out.println(colorize("Unknown command: " + cmd, RED));
        }
    }
}
